Fix checkfail in raw_ops.DenseBincount

SuryanarayanaY opened the pull request:
The API `raw_ops.DenseBincount` lacks validation of input to be vector. It does have checking for rank<=2 but not check for rank>0. Passing a scalar value causes checkfail with debug build.

Reported at #63068

SuryanarayanaY committed:
Fix checkfail in DenseBincount

The API raw_ops.DenseBincount lacks validation of input to be vector. It does have checking for rank<=2 but not for rank>0.
Passing a scalar value causes checkfail with debug build.

Reported at #63068

cantonios commented on tensorflow/core/ops/math_ops.cc:
```diff
@@ -1896,6 +1896,9 @@ REGISTER_OP("DenseBincount")
       } else if (c->Rank(c->input(0)) == 2) {
         c->set_output(0, c->MakeShape({c->Dim(c->input(0), 0), size_val}));
+      } else {
+        return errors::InvalidArgument("input must not be a scalar. "
```
Shouldn't you check if the rank is exactly 0 for scalars?  Also, I think this will fail for dynamic ranks, which will return -1 since the rank is unknown.

SuryanarayanaY replied:
Agree. I missed the case of dynamic inputs.It's better to be explicit about scalar rank.

cantonios replied:
Maybe this should have been my first question: are you sure it doesn't support scalars?  The docs don't say this explicitly.  Should be easy enough to get working for scalars - just treat it as a 1D tensor of size 1.

SuryanarayanaY replied:
Actually I referred this [raw_ops.DenseBincount](https://www.tensorflow.org/api_docs/python/tf/raw_ops/DenseBincount) document


Args
--
input | A Tensor. Must be one of the following types: int32, int64. 1D or 2D int Tensor.

The high level API [tf.math.bincount](https://www.tensorflow.org/api_docs/python/tf/math/bincount) seems accepting scalar also.

SuryanarayanaY replied:
> Maybe this should have been my first question: are you sure it doesn't support scalars? The docs don't say this explicitly. Should be easy enough to get working for scalars - just treat it as a 1D tensor of size 1.

Agree.The raw_ops.DenseBinCount documentation confused me. The checkfail happening due to missing of Shape setting in SetShapeFn with input of rank 0 which is being verified below. 
https://github.com/tensorflow/tensorflow/blob/83f1804f3427ae888e62b26b5bcba8afc9e24ef7/tensorflow/core/framework/shape_inference.cc#L111-L115

I think we can set the same shape as with Rank 1 and remove error statement.I checked with high level API and the output is same for scalar and same scalar as 1D tensor.

```
if (c->Rank(c->input(0)) == 1 || c->Rank(c->input(0)) == 0) {
  c->set_output(0, c->MakeShape({size_val}));
```
WDYT?

SuryanarayanaY replied:
Actually High level API converts input scalar into 1D Tensor before passing to C++ Op. [Gist](https://colab.sandbox.google.com/gist/SuryanarayanaY/81683d768404736922d91c91323ae060/63114_r2.ipynb).

I think we need to do same at C++ Op for scalar inputs.

cantonios replied:
Sounds okay to me.

SuryanarayanaY replied:
Done the changes to set Output shape to Scalar values also. But I believe the Op implementation lacks for Scalar input.

https://github.com/tensorflow/tensorflow/blob/ccb7b001c699d9f288e1c8c8879dc586fd72c727/tensorflow/core/kernels/bincount_op.cc#L311

https://github.com/tensorflow/tensorflow/blob/ccb7b001c699d9f288e1c8c8879dc586fd72c727/tensorflow/core/kernels/bincount_op.cc#L324

There is no implementation for data.dims()=0

SuryanarayanaY replied:
Done. Thanks!

mihaimaruseac commented on tensorflow/core/ops/math_ops.cc:
```diff
@@ -1896,6 +1896,9 @@ REGISTER_OP("DenseBincount")
       } else if (c->Rank(c->input(0)) == 2) {
         c->set_output(0, c->MakeShape({c->Dim(c->input(0), 0), size_val}));
+      } else {
+        return errors::InvalidArgument("input must not be a scalar. "
```
Also, please use absl symbols instead of the TF wrapper

SuryanarayanaY replied:
Done the changes as suggested.

SuryanarayanaY committed:
Update math_ops.cc

Change the logic to validate rank != 0 explicitly

mihaimaruseac commented:
This seems reasonable to me, but I'll let @cantonios to give final approval. Thanks for doing these fixes!

SuryanarayanaY committed:
Set output shape to rank 0 input in DensebinCount 

Set output shape to rank 0 input in DensebinCount  Op.

SuryanarayanaY committed:
Support for Rank 0 Input for DenseBinCount Op

Support for Rank 0 Input for DenseBinCount Op. Assuming that tensor.flat works with scalar tensor also.

mihaimaruseac reviewed (approved):
Thank you!
